### ISSUE TITLE
[question/rfe?] systemd-gpt-auto-generator breaks dracut-generated hostonly image

### ISSUE THREAD

**Issue by RomanValov** [systemd/systemd#35628]
### Component

systemd-gpt-auto-generator

### Is your feature request related to a problem? Please describe

I'm using `dracut` utility to generate initrd/initramfs images for my system. When using dracut with `--hostonly --hostonly-cmdline` arguments it's possible to generate completely self-contained image that has all the configuration and code needed to find, mount and boot from a root on a specific system.

In order to do so, dracut stores needed "kernel" command line parameters inside the image in files in `/etc/cmdline.d` directory. In particular it stores parameters like `root=` and friends in there. As a result there is no need to define `root=` kernel parameter on a real kernel command line in case. And strictly speaking dracut (as an initrd/initramfs image generator) has no direct responsibility and possibility to manage real kernel command line.

The problem arises when gpt-auto-generator comes into play. Due to missing `root=` parameter on kernel command line it tries to detect root automatically but fails to do so or conflicts with root device managed by dracut. I have tested on various distros. Results are differ among platforms but generally the boot process is stuck or broken.

### Describe the solution you'd like

It's possible to use `systemd.gpt_auto=0` kernel command line parameter to workaround the issue. But the question is:

is there proposed way to somehow disable gpt-auto-root detection from dracut's initrd code in case where dracut is able to find and mount the root on it's own based on embedded kernel command line parameters?

or maybe there should be implemented a way to provide the functionality?

For more details please refer dracutdevs/dracut#1722 . One of dracut maintainers suggested to discuss the question with systemd maintainers but I wasn't able to find any relevant discussion here. Sorry if I've missed one.

### Describe alternatives you've considered

1) Obviously, using `systemd.gpt_auto=0`. But wanted to avoid dealing with kernel command line. Finally it may be impossible if using dracut generated UKIs and running with SecureBoot enabled. (see also dracut-ng/dracut-ng#1038) and may lead to completely broken boot.

2) Using SystemdOptions EFI variable. But it's considered deprecated and moreover doesn't work with SecureBoot enabled.

3) Using `SYSTEMD_PROC_CMDLINE` and/or `SYSTEMD_EFI_OPTIONS`. Requires hacky ways to pass environment variables to systemd to be running as PID 1. Moreover these variables doesn't look like officially supported systemd interface.

### The systemd version you checked that didn't have the feature you are asking for

256 / 257

**Comment by arvidjaar** [comment]
> disable gpt-auto-root detection from dracut's initrd code

Not really. With `systemd` as initramfs `/init` generators run before any dracut code and so dracut code cannot affect them.

I am pretty sure it came up before already. Supporting additional files as kernel arguments source looks pretty straightforward, all infrastructure is already in place. These files are part of initrd and so will be measured when TPM2 is used.

**Comment by poettering** [comment]
i am not sure i follow what this is about? they want to override the root= parameter fstab-generator processes from userspace? why? what's the usecase? i don't get it? 

if they want to embed root= inside the initrd, they can already do that? just ship an /etc/fstab with an entry for the /sysroot/ mount, and then do root=fstab on the kernel cmdline? but also, why?

**Comment by RomanValov** [comment]
> just ship an /etc/fstab with an entry for the /sysroot/ mount, and then do root=fstab on the kernel cmdline? but also

@poettering thanks for hint to dig into right direction. Looks like using `/sysroot` mount in initrd's fstab is a way to go. 

Indeed it seems `root=` is not really needed in case. At least as of current systemd (v256/v257). It's due to fstab-generator units (/run/generator) effectively override gpt-auto-generator units (/run/systemd/generator.late). Is it okay to rely upon this behavior? Looks like it was introduced by #3551 and #26038.

Upon quick testing however I also bumped in a fact that `systemd-fstab-generator` spits an error if there is `/sysroot` entry in fstab and `root=/dev/...` is specified on kernel command line in the same time:

```
Failed to create unit file '/run/systemd/generator/sysroot.mount', as it already exists. Duplicate entry in '/etc/fstab'?
```

Nevertheless it doesn't prevent system to boot. But it would be nice if it could be fixed.

> but also, why?

I can't say on behalf of dracut developers. But hostonly mode (embedded `root=`) is encouraged by some distros. Personally as a end-user I have bumped into a case when I was setting up FDE+TPM2 on my laptop. And to the contrary on my side it was like: why do I need to duplicate configuration for root device on the kernel command line if it's already managed by initrd?

**Comment by RomanValov** [comment]
After a bit more experiments with sysroot and fstab (within the context of missing root= on kernel command line) I have bumped into a case where using `/sysroot` in fstab weren't sufficient. The case is:

1) A multi-boot drive with multiple (luks) root partitions used by different distributions
2) One of the root partitions (i.e. fedora) uses `SD_GPT_ROOT_X86_64` for its root device, others uses simply `Linux` type.
3) When attempted to boot into distro with `Linux`-type root (i.e. debian) both, systemd-fstab-generator and systemd-gpt-auto-generator produce valid but independent sysroot mounts. The mount for systemd-gpt-auto-generator is rendered as `-.mount` in systemd state.
4) The boot is not broken in case but mount generated by gpt-auto-generator still invoked and may ask passphrase, thus preventing smooth headless boot expirience.

Finally I have found that dracut has it's own generator and have similar behavior to fstab:/sysroot handled by `systemd-fstab-generator`. Whilst support for fstab:/sysroot is not implemented by dracut at the moment, dracut-rootfs-generator runs some logic in case when `root=` is not found on kernel command line but is known from embedded configuration. In my experiments I've added creation of symlinks to /dev/null aimed to suppress dev-gpt\\x2dauto\\x2droot.device and dev-gpt\\x2dauto\\x2droot\\x2dluks.device by dracut-rootfs-generator. And it seems to works in cases I could think of.

Could you guys please take a look at the PR and share your opinion if it's a valid approach to use? The PR is on dracut-ng repo: https://github.com/dracut-ng/dracut-ng/pull/1063
